rasa_core/policies/embedding_policy.py

```diff
@@ -0,0 +1,848 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import io
+import os
+import logging
+import warnings
+import typing
+from tqdm import tqdm
+
+from typing import \
+    Any, List, Optional, Text
+
+import numpy as np
+from rasa_core.policies import Policy
+from rasa_core.featurizers import \
+    TrackerFeaturizer, FullDialogueTrackerFeaturizer, \
+    LabelTokenizerSingleStateFeaturizer
+
+if typing.TYPE_CHECKING:
+    from rasa_core.domain import Domain
+    from rasa_core.trackers import DialogueStateTracker
+
+try:
+    import cPickle as pickle
+except ImportError:
+    import pickle
+
+logger = logging.getLogger(__name__)
+
+
+try:
+    import tensorflow as tf
+except ImportError:
+    tf = None
+
+
+class EmbeddingPolicy(Policy):
+    SUPPORTS_ONLINE_TRAINING = True
+
+    @classmethod
+    def _standard_featurizer(cls):
+        return FullDialogueTrackerFeaturizer(
+                    LabelTokenizerSingleStateFeaturizer())
+
+    config = {
+        # nn architecture
+        "num_hidden_layers_a": 0,
+        "hidden_layer_size_a": [],
+        "num_hidden_layers_b": 0,
+        "hidden_layer_size_b": [],
+        "rnn_size": 32,
+        "batch_size": 16,
+        "epochs": 3000,
+
+        # embedding parameters
+        "embed_dim": 10,
+        "mu_pos": 0.8,  # should be 0.0 < ... < 1.0 for 'cosine'
+        "mu_neg": -0.2,  # should be -1.0 < ... < 1.0 for 'cosine'
+        "similarity_type": 'cosine',  # string 'cosine' or 'inner'
+        "num_neg": 20,
+        "use_max_sim_neg": True,  # flag which loss function to use
+
+        # regularization
+        "C2": 0.001,
+        "C_emb": 0.8,
+
+        "droprate_a": 0.1,
+        "droprate_b": 0.1,
+        "droprate_c": 0.3,
+        "droprate_rnn": 0.1,
+        "droprate_out": 0.1,
+    }
+
+    def _load_nn_architecture_params(self):
+        self.num_hidden_layers_a = self.config['num_hidden_layers_a']
```

Comment:
I am not quite sure, but what is the benefit of setting all of these as single attributes?

Reply:
no I mean in general, store the whole config `self.config = config` and where needed use `self.config['num_hidden_layers_a']` instead of `self.num_hidden_layers_a` (avoiding all the attributes)

Reply:
I don't have a strong opinion about this, we can also keep it that way (but then `config` shouldn't be an object property as it will duplicate the values and they will run out of sync).

Reply:
tbh, I just copied it from nlu embedding classifier